module-system/module-implementations/sov-nft-module/src/nft.rs

```diff
@@ -0,0 +1,177 @@
+use anyhow::{anyhow, bail};
+use sov_modules_api::{Context, StateMap, WorkingSet};
+
+use crate::collection::Collection;
+use crate::{CollectionAddress, OwnerAddress, UserAddress};
+
+/// tokenId for the NFT that's unique within the scope of the collection
+pub type TokenId = u64;
+
+#[cfg_attr(
+    feature = "native",
+    derive(serde::Serialize),
+    derive(serde::Deserialize)
+)]
+#[derive(borsh::BorshDeserialize, borsh::BorshSerialize, Clone, Debug, PartialEq, Eq, Hash)]
+/// A simple wrapper struct to mark an NFT identifier as a combination of
+/// a token id (u64) and a collection address
+pub struct NftIdentifier<C: Context>(pub TokenId, pub CollectionAddress<C>);
+
+#[cfg_attr(
+    feature = "native",
+    derive(serde::Serialize),
+    derive(serde::Deserialize)
+)]
+#[derive(borsh::BorshDeserialize, borsh::BorshSerialize, Debug, PartialEq, Clone)]
+/// Defines an nft
+pub struct Nft<C: Context> {
+    /// A token id that uniquely identifies an NFT within the scope of a (collection name, creator)
+    token_id: TokenId,
+    /// A collection address that uniquely identifies a collection - derived from (collection name, creator)
+    collection_address: CollectionAddress<C>,
+    /// Owner address of a specific token_id within a collection
+    owner: OwnerAddress<C>,
+    /// A frozen NFT cannot have its data altered and is immutable
+    /// Cannot be unfrozen. token_uri cannot be modified
+    frozen: bool,
+    /// A URI pointing to the offchain metadata
+    token_uri: String,
+}
+
+/// NewType representing an owned NFT
+/// An owned NFT is owned by the context sender and is transferable
+pub struct OwnedNft<C: Context>(pub Nft<C>);
+
+/// NewType representing a Mutable NFT
+/// A mutable NFT is modifiable by the creator, but only certain fields (frozen, token_uri)
+pub struct MutableNft<C: Context>(pub Nft<C>);
+
+impl<C: Context> OwnedNft<C> {
+    pub fn set_owner(&mut self, to: &UserAddress<C>) {
+        self.0.owner = OwnerAddress::new(to.get_address())
+    }
+}
+
+impl<C: Context> MutableNft<C> {
+    pub fn freeze(&mut self) {
+        self.0.frozen = true;
+    }
+    pub fn update_token_uri(&mut self, token_uri: &str) {
+        self.0.token_uri = token_uri.to_string();
+    }
+}
+
+impl<C: Context> Nft<C> {
+    pub fn new(
+        token_id: TokenId,
+        token_uri: &str,
+        mint_to_address: &UserAddress<C>,
+        frozen: bool,
+        collection_address: &CollectionAddress<C>,
+        nfts: &StateMap<NftIdentifier<C>, Nft<C>>,
+        working_set: &mut WorkingSet<C>,
+    ) -> anyhow::Result<Self> {
+        if nfts
+            .get(
+                &NftIdentifier(token_id, collection_address.clone()),
+                working_set,
+            )
+            .is_some()
+        {
+            bail!(
+                "NFT with id {} already exists for collection address {}",
+                token_id,
+                collection_address
+            )
+        }
```

Comment:
nit: the anyhow `Context` trait provides a nice method for this. `with_context` is like expect, but bails instead of panicking.